### ISSUE TITLE
Gradients are being updated on the augmented data, according to the paper this shouldn't be the case

### ISSUE THREAD

**Issue by preachermaan** [ildoonet/unsupervised-data-augmentation#7]
According to the paper:

> Specifically,following VAT [44], we choose to minimize the KL divergence between the predicted distributions on an unlabeled example and an augmented unlabeled example:
> 
> minθ JUDA(θ) = E x∈U E xˆ∼q(ˆx|x) DKL pθ˜(y | x) pθ(y | xˆ)) , (1)
> 
> where q(ˆx | x) is a data augmentation transformation and ˜θ is a fixed copy of the current parameters θ indicating that the gradient is not propagated through ˜θ, as suggested by Miyato et al

So basically, when we do a forward pass on the model on unlabel2 we need to do with torch.no_grad() so we don't update the gradients for the forward pass on the augmented data.

I also _think_ that we might want to do model.eval() for both unlabel1 and unlabel2 to turn off dropout (if any is being used) and to stop batch norm statistics being updated. Not 100% sure on this though.

**Comment by preachermaan** [comment]
Actually, I think I have that the wrong way round. Gradients should be propagated through the augmented but not the unlabeled augmented examples. Either way, we definitely should not be doing both.

**Comment by ildoonet** [comment]
@safekidda Do you mean this? : https://github.com/ildoonet/unsupervised-data-augmentation/blob/master/train.py#L62

I detached unlabel data tensor, where detaching means no gradient will be calculated on the tensor.

**Comment by preachermaan** [comment]
Thanks, I didn't spot this. I also wasn't aware that detach exhibited this behavior.

With regards my other point on model.eval() to disable dropout and batch norm, I've been in touch with one of the authors of the paper on Reddit, and he ran experiments with both and it was better leaving any regularizer and batch norm updates in place, so I'll close this issue.

**Comment by ildoonet** [comment]
Thanks. I also have a same result of batch norm disabling on unlabel data.

**Comment by preachermaan** [comment]
Okay, you've tested it? Was there much difference?

**Comment by ildoonet** [comment]
I got slightly worse result when disabling bn.
